src/transformers/models/owlvit/modeling_owlvit.py

```diff
@@ -0,0 +1,1400 @@
+        query_embeds /= torch.linalg.norm(query_embeds, dim=-1, keepdim=True) + 1e-6
+
+        # Get class predictions
+        pred_logits = torch.einsum("...pd,...qd->...pq", image_class_embeds, query_embeds)
+
+        # Apply a learnable shift and scale to logits
+        logit_shift = self.logit_shift(image_embeds)
+        logit_scale = self.logit_scale(image_embeds)
+        logit_scale = self.elu(logit_scale) + 1
+        pred_logits = (pred_logits + logit_shift) * logit_scale
+
+        if query_mask is not None:
+            if query_mask.ndim > 1:
+                query_mask = torch.unsqueeze(query_mask, dim=-2)
+
+            pred_logits = pred_logits.to(torch.float64)
+            pred_logits = torch.where(query_mask == 0, -1e6, pred_logits)
+
+        return (pred_logits, image_class_embeds)
+
+
+class OwlViTImageTextEmbedder(nn.Module):
+    def __init__(self, config: OwlViTConfig):
+        super().__init__()
+
+        self.clip = OwlViTModel(config)
+        self.layer_norm = nn.LayerNorm(config.vision_config.hidden_size)
+
+    def forward(
+        self,
+        pixel_values: Optional[torch.FloatTensor] = None,
+        input_ids: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+
+        image_embeds, text_embeds = None, None
+
+        # Encode text
+        if input_ids is not None:
+            text_embeds = self.clip.get_text_features(input_ids=input_ids, attention_mask=attention_mask)
+
+        # Encode image
+        if pixel_values is not None:
+            image_embeds = self.clip.get_image_features(pixel_values, return_projected=False)
+
+            # Resize class token
+            new_size = tuple(np.array(image_embeds.shape) - np.array((0, 1, 0)))
+            class_token_out = torch.broadcast_to(image_embeds[:, :1, :], new_size)
+
+            # Merge image embedding with class tokens
+            image_embeds = image_embeds[:, 1:, :] * class_token_out
+            image_embeds = self.layer_norm(image_embeds)
+
+        return (image_embeds, text_embeds)
+
+
+class OwlViTForObjectDetection(OwlViTPreTrainedModel):
+    def __init__(self, config: OwlViTConfig):
+        super().__init__(config)
+
+        self._embedder = OwlViTImageTextEmbedder(config)
+        self._class_head = OwlViTClassPredictionHead(config)
+        self._box_head = OwlViTBoxPredictionHead(config)
+        self.sigmoid = nn.Sigmoid()
+
+    def normalize_grid_corner_coordinates(self, feature_map: torch.FloatTensor):
+        # Computes normalized xy corner coordinates from feature_map.
+        if not feature_map.ndim == 4:
+            raise ValueError("Expected input shape is [batch_size, num_channels, height, width]")
+
+        h, w = feature_map.shape[1:3]
+
+        xy = np.stack(np.meshgrid(np.arange(1, w + 1), np.arange(1, h + 1)), axis=-1).astype(np.float32)
```

Comment:
Please use better names here :-) `height`, `width` and I have no idea what `xy` means :D

Reply:
Yes, very good point. I'll change it to make all variable names more self-descriptive